rfcs/29_centralized_run_criteria.md

```diff
@@ -0,0 +1,96 @@
+
+## User-facing explanation
+
+Run criteria are utility systems that control whether regular systems should run during a given schedule execution. They can be used by entire schedules and stages, or by individual systems. Any system that returns `ShouldRun` can be used as a run criteria.
+
+Criteria can be defined in two ways:
+- Inline, directly attached to their user:
+```rust
+fn run_if_player_has_control(controller: Res<PlayerController>) -> ShouldRun {
+    if controller.player_has_control() {
+        ShouldRun::Yes
+    } else {
+        ShouldRun::No
+    }
+}
+
+stage
+    .add_system(player_movement_system.with_run_criteria(run_if_player_has_control))
+    .add_system(
+        player_attack_system.with_run_criteria(|controller: Res<PlayerController>| {
+            if controller.player_can_attack() {
+                ShouldRun::Yes
+            } else {
+                ShouldRun::No
+            }
+        }),
+    );
+```
+- Globally, owned by a schedule, usable by anything inside said schedule via a label:
+```rust
+#[derive(Debug, Clone, PartialEq, Eq, Hash, RunCriteriaLabel)]
+enum PlayerControlCriteria {
+    CanAttack,
+}
+
+schedule
+    .add_run_criteria(run_if_player_can_attack.label(PlayerControlCriteria::CanAttack))
+    .add_system(player_attack_system.with_run_criteria(PlayerControlCriteria::CanAttack))
+    .add_system(player_dmg_buff_system.with_run_criteria(PlayerControlCriteria::CanAttack))
+```
+
+Note that only labels that appear on just one criteria can be utilized for reusing criteria results: the label to criteria relationship must be unambiguous.
+
+Since run criteria aren't forbidden from having side effects, it might become necessary to define an explicit execution order between them. This is done exactly as with regular systems:
+```rust
+schedule
+    .add_run_criteria(run_if_player_can_attack.label(PlayerControlCriteria::CanAttack))
+    .add_run_criteria(run_if_player_can_move.after(PlayerControlCriteria::CanAttack))
+```
+
+It should be noted that criteria declared globally are required to have a label, because they would be useless for their purpose otherwise. Criteria declared inline can be labelled as well, but using that for anything other than defining execution order is detrimental to code readability.
+
+Run criteria are evaluated once at the start of schedule, any criteria returning `ShouldRun::*AndCheckAgain` will be evaluated again at the end of schedule. At that point, depending on the new result, the systems governed by those criteria might be ran again, followed by another criteria evaluation; this repeats until no criteria return `ShouldRun::Yes*`. Systems with no run criteria always run exactly once per schedule execution.
```

Comment:
Does this mean that a schedule with the following stages:

UpdateStage: MoveSystem (fixed timestep criteria)
PostUpdateStage: TransformPropagateSystem
RenderStage: RenderSystem

Could execute like this for a single frame?

MoveSystem (Yes and Loop)
TransformPropagateSystem
RenderSystem
MoveSystem (No)

Isn't this, like, very wrong? MoveSystem is intended to be run X times for the fixed timestep prior to Transforms being updated and RenderSystem rendering the final transform state. Instead, only the first timestep is rendered, then subsequent timesteps happen after rendering and show up next frame (and note that the position would be out of sync with the TransformPropagate state at the start of the next frame). Am I confused?

Reply:
No, you're right, however: this feature is breaking; what you outlined is a prime example of the breakage. The schedule and criteria that's okay in current implementation will most likely become nonsense - I should probably put that in drawbacks.

The intended solution is to either involve everything in a FSM, or rearrange the schedule to have all systems like "RenderSystem" at the start of the schedule. I'm not absolutely confident that's a good solution (on its own, anyway), hence the RFC.

Reply:
> The intended solution is to either involve everything in a FSM, or rearrange the schedule to have all systems like "RenderSystem" at the start of the schedule.

I don't find the "move render system to the front" approach very satisfying. Render _shouldn't_ run until PostUpdate runs, otherwise it will attempt to draw stuff that hasn't been "fully processed" yet. Putting it first feels wrong. We would need to create new machinery to ensure that doesn't happen on the first run, which feels complicated to me, especially if there are multiple stages that all rely on each other (ex: a PostRender stage). And it would mean that rendering happens for "the previous update" instead of "the current update". And managing "dependencies" between multiple stages (instead of just the Update->Render dependency) when they aren't necessarily inserted in the order they should run sounds nasty.

FSM might be a solution, but it feels like we should wait for that design before committing to this impl.